### PR TITLE
rouille -> warp

### LINKED ISSUE #39: move away from ` rouille`
Currently `rouille` is used which is heavily thread based, there is an upstream effort to move to `actix-web`, but I am not sure this is what we want. https://github.com/mozilla/sccache/issues/742#issuecomment-808824589

Evaluate alternative crates to replace `rouille`:

* [ ] `actix-web`
* [ ] `hyper`
* [ ] `tide`
* [ ] ...

### PR DESCRIPTION
**Pull request by montekki**
Still a lot of work to do but I think that api will be compatible with old code.

Fixes #39

### CODE REVIEW

**Comment by drahnr** [comment]
I'll give this a spin at some point this week, other than that this is looking ready to be merged 🎉

**Comment by drahnr** [comment]
`cachepot -s`

yields

```
cachepot: error: failed to get stats from server
cachepot: caused by: Failed to send data to or receive data from server
cachepot: caused by: Failed to read response header
cachepot: caused by: failed to fill whole buffer
```

**Comment by montekki** [comment]
Were all previously launched instances of daemonized `cachepot` killed beforehand?

**Comment by drahnr** [comment]
> Were all previously launched instances of daemonized `cachepot` killed beforehand?

Double checked:

After running `sudo systemctl restart cachepot-server.service cachepot-scheduler.service; sudo pkill cachepot; ps ax | rg cachepot` with only the `rg` process itself, same result.

**Comment by montekki** [comment]
Damn, I cant repro and i didn't even touch `src/server.rs` (i think)

**Comment by montekki** [comment]
Can you `RUST_LOG=trace`?

**Comment by montekki** [comment]
Also, I've run into `ufw` blocking connects

**Comment by drahnr** [comment]
I think my local redis just crashed and this is a side effect of it. I'll try to fix the instance, and open an issue regarding error handling of the backend if this is resolved.

**Comment by Xanewok** [comment]
> `cachepot -s`
> 
> yields
> 
> ```
> <error>
> ```

I can't repro, I'm getting regular stat dump using this branch

**Comment by drahnr** [comment]
After fixing `redis`, the output is as expected.

File: src/util.rs [hunk 1]
```diff
@@ -544,6 +531,29 @@ pub fn daemonize() -> Result<()> {
...
+{
+    let mut tls_builder = native_tls::TlsConnector::builder();
+
+    for root_cert in root_certs {
+        tls_builder.add_root_certificate(native_tls::Certificate::from_pem(root_cert.as_ref())?);
+    }
+
+    tls_builder.use_sni(false);
+
+    let tls = tls_builder.build()?;
```
Comment (drahnr) [inline src/util.rs:548]: A few comments on this would be appreciated.

**Review by drahnr — APPROVED** [review]

**Review by Xanewok — COMMENTED** [review]
Excited for this change! 🎉 Looks good at a glance, I'll dig deeper later on since I had a hard time following the logic bundled together in a single file

File: src/util.rs (hunk 1 above)
Comment (montekki) [reply src/util.rs:561]: done

File: src/dist/http.rs [hunk 2]
```diff
@@ -651,80 +444,891 @@ mod server {
...
+                http::server::{ClientVisibleMsg, ErrJson},
+                JobAuthorizer, JobId, ServerIncoming,
+            };
+
+            fn bearer_http_auth(auth_header: &HeaderValue) -> Result<String, Error> {
+                let header = auth_header
+                    .to_str()
+                    .map_err(|_| Error::AuthorizationHeaderBroken)?;
+
+                let mut split = header.splitn(2, |c| c == ' ');
```
[the remark this answers is not in the document]
Comment (Xanewok) [reply src/dist/http.rs:490]: Ah, makes sense then!

File: src/dist/http.rs [hunk 3]
```diff
@@ -651,80 +444,891 @@ mod server {
...
+                    .handle_run_job(requester.as_ref(), job_id, command, outputs, inputs_rdr)
+                    .await
+                    .map_err(|_| warp::reject::custom(Error::AssignJob))?;
+
+                Ok(res)
+            }
+        }
+    }
+
+    mod scheduler_api_v1 {
```
[the remark this answers is not in the document]
Comment (montekki) [reply src/dist/http.rs:812]: I would suggest in another changeset.

**Comment by montekki** [comment]
Merged this since it is blocking all stuff that needs to be done on top of it. In case there (most likely) be fallout lets deal with it iteratively in new issues.

**Pull request merged**